File: LeetCode/47全排列Ⅱ.go
var ans [][]int
var path []int

func permuteUnique(nums []int) [][]int {
	ans = [][]int{}
	path = []int{}
	used := make([]bool, len(nums))
	sort.Ints(nums)
	backtracking(nums, used)
	return ans
}

func backtracking(nums []int, used []bool) {
	if len(path) == len(nums) {
		temp := make([]int, len(nums))
		copy(temp, path)
		ans = append(ans, temp)
	}
	for i := 0; i < len(nums); i++ {
		if used[i] == true || i > 0 && nums[i] == nums[i-1] && used[i-1] == false {
			continue
		}
		used[i] = true
		path = append(path, nums[i])
		backtracking(nums, used)
		path = path[:len(path)-1]
		used[i] = false
	}
}

File: LeetCode/47全排列Ⅱ.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> permuteUnique(int[] nums) {
        boolean[] used = new boolean[nums.length];
        Arrays.sort(nums);
        Arrays.fill(used, false);
        backtracking(nums, used);
        return ans;
    }

    public void backtracking(int[] nums, boolean[] used) {
        if (path.size() == nums.length) {
            ans.add(new ArrayList<>(path));
            return ;
        }
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1]) continue;
            if (used[i]) continue;
            used[i] = true;
            path.add(nums[i]);
            backtracking(nums, used);
            path.removeLast();
            used[i] = false;
        }
    }
}